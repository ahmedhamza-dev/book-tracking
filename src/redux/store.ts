import { configureStore } from "@reduxjs/toolkit";
import bookSlice from "./slices";
import createSagaMiddleware from 'redux-saga'
import saga from "./saga";

export const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
    reducer: {
        allBooks: bookSlice.reducer
    },
    middleware:[sagaMiddleware]
});
sagaMiddleware.run(saga);

export default store;