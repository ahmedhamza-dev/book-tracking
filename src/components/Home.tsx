import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import Shelf from "./Shelf";
import { getAllBooks } from "../redux/slices";

const Home = () => {
  const state: any = useSelector((state) => state);
  const dispatch = useDispatch<any>();
  const { allBooks } = state;
  useEffect(
    () => {
      dispatch(getAllBooks());
    },
    [dispatch]
  );
  return (
    <div className="list-books">
      <div className="list-books-title">
        <h1>MyReads</h1>
      </div>
      <div className="list-books-content" data-testid="books-content">
        <div>
          {allBooks.books && (
            <>
              <Shelf
                data-testid="currently-Reading"
                title="Currently Reading"
                category="currentlyReading"
                books={allBooks.books}
              />
              <Shelf
                data-testid="wantToRead"
                title="Want to Read"
                category="wantToRead"
                books={allBooks.books}
              />
              <Shelf
                data-testid="Read"
                title="Read"
                category="read"
                books={allBooks.books}
              />
            </>
          )}
        </div>
      </div>
      <div className="open-search">
        <Link to="/search" className="open-search-link">
          Add a Book
        </Link>
      </div>
    </div>
  );
};

export default Home;
