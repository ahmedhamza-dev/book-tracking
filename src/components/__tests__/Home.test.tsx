import React from "react";
import "@testing-library/jest-dom";
import Home from "../Home";
import renderer from "react-test-renderer";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { BrowserRouter as Router, Switch } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../../redux/store";

afterEach(() => {
  cleanup();
});

it("Home components testing", () => {
  const homeComponent = (
    <>
      <Provider store={store}>
        <Router>
          <Switch>
            <Home />
          </Switch>
        </Router>
      </Provider>
    </>
  );
  const component = renderer.create(homeComponent);
  const { getByTestId, getByText } = render(homeComponent);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const booksContent = getByTestId("books-content");
  expect(booksContent).toBeInTheDocument();

  const link = getByText("Add a Book");
  expect(window.location.pathname).toBe("/");
  fireEvent.click(link);
  expect(window.location.pathname).toBe("/search");
});
