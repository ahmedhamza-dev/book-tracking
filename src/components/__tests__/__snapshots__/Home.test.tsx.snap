// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home components testing 1`] = `
Array [
  <h2
    className="sc-bcXHqe icDmVn"
  >
    MyReads
  </h2>,
  <div
    className="sc-gswNZR jOpWbm"
    data-testid="books-content"
  >
    <div
      className="sc-hLBbgP inROHo"
    >
      <h2
        className="sc-eDvSVe jSzroE"
        data-testid="shelf-title"
      >
        Currently Reading
      </h2>
      <div
        className="sc-jSUZER cUjmio"
      >
        <div
          className="sc-gKPRtg hzXScj"
          data-testid="books-grid"
        />
      </div>
    </div>
    <div
      className="sc-hLBbgP inROHo"
    >
      <h2
        className="sc-eDvSVe jSzroE"
        data-testid="shelf-title"
      >
        Want to Read
      </h2>
      <div
        className="sc-jSUZER cUjmio"
      >
        <div
          className="sc-gKPRtg hzXScj"
          data-testid="books-grid"
        />
      </div>
    </div>
    <div
      className="sc-hLBbgP inROHo"
    >
      <h2
        className="sc-eDvSVe jSzroE"
        data-testid="shelf-title"
      >
        Read
      </h2>
      <div
        className="sc-jSUZER cUjmio"
      >
        <div
          className="sc-gKPRtg hzXScj"
          data-testid="books-grid"
        />
      </div>
    </div>
  </div>,
  <div
    className="sc-dkrFOg iaXMdo"
  >
    <a
      className="open-search-link"
      href="/search"
      onClick={[Function]}
    >
      Add a Book
    </a>
  </div>,
]
`;
